File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        // mark each current element
        // for each element, iterate using two pointers moving them concurrently
        // if the current sum with left + right + current < 0, move the left pointer by 1
        // if > 0, move the right pointer by 1
        // otherwise, add it to the list

        // to check for duplicates, if we add something we iterate until we the first distinct element for the left and right pointers
        // otherwise, for the current element, we check if we have adjacent elements and continue
        List<List<Integer>> ret = new ArrayList<>();

        for(int i = 0; i < nums.length - 2; i++) {
            if(i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1;
            int right = nums.length - 1;

            while(left < right) {
                int sum = nums[left] + nums[right] + nums[i];

                if(sum == 0) {
                    ret.add(Arrays.asList(nums[left], nums[right], nums[i]));

                    while(left < right && nums[left] == nums[left + 1]) left++;
                    while(left < right && nums[right] == nums[right - 1]) right--;

                    left++;
                    right--;
                } else if(sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return ret;
    }
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        # have a two pointer solution, reduce from O(n^3) -> O(n^2)
        # basically, every unique value starting from index 1 and going to n - 1 (exclusive)
        # have a left pointer and a right pointer at the edges 
        # move it together based on the rules of just adding
        # make sure to sort first

        # [-1, -1, -1, 0, 1, 2]
        # 

        nums.sort()

        ret = []

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            left = i + 1
            right = len(nums) - 1

            while left < right:
                s = nums[left] + nums[right] + nums[i]
                if s < 0:
                    left += 1
                elif s > 0:
                    right -= 1
                else:
                    ret.append([nums[left], nums[right], nums[i]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left - 1]: left += 1
                    while left < right and nums[right] == nums[right + 1]: right -= 1
        return ret

                
                    
            